clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java

```diff
@@ -346,4 +347,85 @@ public AlterConfigsResult alterConfigs(Map<ConfigResource, Config> configs) {
      * @return                The AlterConfigsResult
      */
     public abstract AlterConfigsResult alterConfigs(Map<ConfigResource, Config> configs, AlterConfigsOptions options);
+
+    /**
+     * Change the log directory for the specified replicas. This API is currently only useful if it is used
+     * before the replica has been created on the broker. It will support moving replicas that are already created after
+     * KIP-113 is fully implemented.
+     *
+     * Updates are not transactional so they may succeed for some resources while fail for others. The log directory for
+     * a particular replica is updated atomically.
+     *
+     * This operation is supported by brokers with version 0.11.1.0 or higher.
+     *
+     * @param replicaAssignment  The replicas with their log directory absolute path
+     * @return                   The AlterReplicaDirResult
+     */
+    public AlterReplicaDirResult alterReplicaDir(Map<TopicPartitionReplica, String> replicaAssignment) {
```

Comment:
@becketqin Thanks for your comment! Here is what I think:

- I think the interface is useful. The AlterReplicaDirRequest (renamed from ChangeReplicaDirRequest per Ismael's comment) is used to 1) create replica in the specified log directory later and 2) move replica to the specified log directory if it has already been created yet. This patch supports the first part and thus the new API `alterReplicaDir` in AdminClient is useful and well-defined. If we don't add this method and its implementation in AdminClient, we would have to implement it in `ReassingPartitionCommand` and move the implementation to AdminClient later, which seems like unnecessary work.

- Can you clarify a bit what implementation detail is exposed to user in this patch?

- I have thought about the possibility of using only one method in AdminClient to do both partition reassignment and replica -> log directory reassignment. My conclusion is that it is still better to put them into two separate methods. For example, if we do both using one method, the method would probably look like `reassignPartition(Map<TopicPartition, List<Int>> newPartitionAssignment, Map<TopicPartitionReplica, String> newReplicaDirAssignment)`. This method merges two inherently different parameters and functionality into one method, which seems less clean than using two methods. Do you any suggestion on how that one method would look like if we were to use only one method for this?

- I agree with the tasks 1-3. I still think it is better to put `alterReplicaDir` in AdmClient as long as we documents it properly. But it also works for me to move this code to `ReassignPartitionCommand`.